lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -233,9 +261,470 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  }
+  // edgesToFix also has edges to the header that are not exit edges.
+  // To simplify the code below, simply assign an arbitrary value (say 0)
+  // as the index for header. This would get simplified once we unify the
+  // stayInloop flag and exitIndex into one value.
+  exitIndices.insert({header, 0});
+  unsigned nextExitIndex = 1;
+  for (const auto &edge : edgesToFix) {
+    SILBasicBlock *src = const_cast<SILBasicBlock *>(edge.first);
+    SILBasicBlock *tgt = const_cast<SILBasicBlock *>(edge.second);
+    SILBasicBlock *newTgt = (src == preheader) ? newHeader : latchBlock;
+    bool stayInLoop = loop->contains(tgt);
+    replaceBranchTarget(src->getTerminator(), tgt, newTgt,
+                        /*preserveArgs=*/stayInLoop);
+    // Set up additional arguments.
+    // If we are exiting the loop, then we should simply use newHeader args.
+    SmallVector<SILValue, 8> newArgs;
+    if (!stayInLoop) {
+      for (unsigned i = 0; i < oldHeaderNumArgs; ++i) {
+        newArgs.push_back(newHeader->getArgument(i));
+      }
+    }
+    SILBuilder builder(src->getTerminator());
+    SILLocation location(
+        getUserSourceLocation(src->getTerminator()->getDebugLocation()));
+    // Find an appropriate value to use for each escaping value.
+    unsigned argIndex = oldHeaderNumArgs;
+    for (const SILValue &escapingValue : escapingValues) {
+      if (DI->properlyDominates(escapingValue, src->getTerminator())) {
+        newArgs.push_back(escapingValue);
+      } else if (src != preheader) {
+        // newHeader arguments are available if src is not the preheader.
+        newArgs.push_back(newHeader->getArgument(argIndex));
+      } else {
+        // State from within the loop is not available in the preheader.
+        // Simply pass in an undef. This will never be accessed at runtime.
+        newArgs.push_back(getUndef(escapingValue->getType()));
+      }
+      ++argIndex;
+    }
+    // `exitIndex` to identify the block to which we exit from the loop.
+    // (insert a new value or get the old key value pair.)
+    auto emplaceResult = exitIndices.try_emplace(tgt, nextExitIndex);
+    if (emplaceResult.second) {
+      // Increment index as we inserted a new entry into the table.
+      ++nextExitIndex;
+    }
+    auto kvPair = *emplaceResult.first;
+    newArgs.push_back(createTFIntegerConst(builder, location,
+                                           /*bitwidth*/ 32,
+                                           /*exitIndex*/ kvPair.second));
+    // `stayInLoop` flag
+    newArgs.push_back(createTFIntegerConst(builder, location,
+                                           /*bitwidth*/ 1, stayInLoop));
+    appendArguments(src->getTerminator(), newTgt, newArgs);
+  }
+  return exitIndices;
+}
+
+/// Build a demuxing if..then..else block that can be used as a single
+/// exit block. This will not have any effect if there is a single exit
+/// block already.
+SILBasicBlock *SingleExitLoopTransformer::createNewExitBlock(
```

Comment:
we are creating more than a single exit block -- maybe call this function sth like `createNewExitBlockWithDemux()`?